cpp/src_prims/sparse/batched/csr.h

```diff
@@ -0,0 +1,721 @@
+
+  if (bid < batch_size) {
+    int stride = m * n;
+    for (int i = 0; i < m; i++) {
+      for (int idx = row_index[i]; idx < row_index[i + 1]; idx++) {
+        int j = col_index[idx];
+        dense[bid * stride + j * m + i] = values[bid * nnz + idx];
+      }
+    }
+  }
+}
+
+/**
+ * @brief The Batched::CSR class provides storage and a few operations for
+ *        a batch of matrices in Compressed Sparse Row representation, that
+ *        share a common structure (index arrays) but different values.
+ */
+template <typename T>
+class CSR {
+ public:
+  /**
+   * @brief Constructor that leaves the matrix uninitialized
+   * 
+   * @param[in] m                Number of rows per matrix
+   * @param[in] n                Number of columns per matrix
+   * @param[in] nnz              Number of non-zero elements per matrix
+   * @param[in] batch_size       Number of matrices in the batch
+   * @param[in] cublasHandle     cuBLAS handle
+   * @param[in] cusolverSpHandle cuSOLVER sparse handle
+   * @param[in] allocator        Device memory allocator
+   * @param[in] stream           CUDA stream
+   */
+  CSR(int m, int n, int nnz, int batch_size, cublasHandle_t cublasHandle,
+      cusolverSpHandle_t cusolverSpHandle,
+      std::shared_ptr<ML::deviceAllocator> allocator, cudaStream_t stream)
+    : m_batch_size(batch_size),
+      m_allocator(allocator),
+      m_cublasHandle(cublasHandle),
+      m_cusolverSpHandle(cusolverSpHandle),
+      m_stream(stream),
+      m_shape(std::pair<int, int>(m, n)),
+      m_nnz(nnz) {
+    // Allocate the values
+    T* values =
+      (T*)m_allocator->allocate(sizeof(T) * m_nnz * m_batch_size, m_stream);
+    // Allocate the index arrays
+    int* col_index = (int*)m_allocator->allocate(sizeof(int) * m_nnz, m_stream);
+    int* row_index =
+      (int*)m_allocator->allocate(sizeof(int) * (m_shape.first + 1), m_stream);
+
+    /* Take these references to extract them from member-storage for the
+     * lambda below. There are better C++14 ways to do this, but I'll keep
+     * it C++11 for now. */
+    auto& shape = m_shape;
+
+    /* Note: we create these "free" functions with explicit copies to ensure
+     * that the deallocate function gets called with the correct values. */
+    auto deallocate_values = [allocator, batch_size, nnz, stream](T* A) {
+      allocator->deallocate(A, batch_size * nnz * sizeof(T), stream);
+    };
+    auto deallocate_col = [allocator, nnz, stream](int* A) {
+      allocator->deallocate(A, sizeof(int) * nnz, stream);
+    };
+    auto deallocate_row = [allocator, shape, stream](int* A) {
+      allocator->deallocate(A, sizeof(int) * (shape.first + 1), stream);
+    };
+
+    // When the shared pointers go to 0, the memory is deallocated
+    m_values = std::shared_ptr<T>(values, deallocate_values);
+    m_col_index = std::shared_ptr<int>(col_index, deallocate_col);
+    m_row_index = std::shared_ptr<int>(row_index, deallocate_row);
```

Comment:
have you looked at `MLCommon::device_buffer`? It was specifically designed to address these kinds of things itself. It can be found defined [here](https://github.com/rapidsai/cuml/blob/branch-0.13/cpp/src_prims/common/device_buffer.hpp).

Reply:
I'll check that, thanks.